p2p: disconnect peers that send filterclear + update existing filter msg disconnect logic

glozow opened the pull request:
Nodes that don't have bloomfilters turned on (i.e. no `NODE_BLOOM` service) should disconnect peers that send them `filterclear` P2P messages.

Non-bloomfilter nodes already disconnect peers for [`filteradd` and `filterload`](https://github.com/bitcoin/bitcoin/blob/19e919217e6d62e3640525e4149de1a4ae04e74f/src/net_processing.cpp#L2218), but #8709 removed `filterclear` so it could be used to reset tx relay. This isn't needed now because using `feefilter` message is much better for this purpose (See #19204).

Also refactors existing disconnect logic for `filteradd` and `filterload` into respective message handlers and removes banning for them.

glozow committed:
[netprocessing] disconnect node that sends filterclear

-nodes not serving bloomfilters should disconnect peers
that send filterclear, just like filteradd and filterload
-nodes that want to enable/disable txrelay should use
feefilter

glozow committed:
[test] test disconnect for filterclear

MarcoFalke commented:
Concept ACK. There shouldn't be a reason for a node to send filterclear when we haven't offered them bloom services to them.

See also BIP 111: 

> If a node does not support bloom filters but receives a "filterload", "filteradd", or "filterclear" message from a peer the node should disconnect that peer immediately

https://github.com/bitcoin/bips/blob/master/bip-0111.mediawiki

nit: I think this is a good opportunity to move the disconnect handling of `NetMsgType::FILTERLOAD` and ` NetMsgType::FILTERADD` into the respective scope of the switch statement instead of having it at the top of the function, disconnected from the actual filter message handling.

glozow committed:
[p2p/refactor] move disconnect logic and remove misbehaving

-Increasing the banscore and/or banning is too harsh,
just disconnecting is enough.
-Return true from ProcessMessage because we already log
receipt of filterclear and disconnect.

gzhao408 commented on src/net_processing.cpp:
```diff
@@ -3447,6 +3433,16 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
     }
 
     if (msg_type == NetMsgType::FILTERLOAD) {
+        if (!(pfrom.GetLocalServices() & NODE_BLOOM)) {
```
This code block is repeated 3 times and bloats the `ProcessMessage` function. Would a helper function like `MaybePunishNodeForBloom` be appropriate or is this fine?

jnewbery replied:
Yes, but I think you can also just reduce the complexity here. I don't think there's any benefit to calling `Misbehaving()` on a peer that sends us a `FILTER*` message. Calling `Misbehaving(100)` results in us disconnecting the peer and then marking it discouraged (see #19219). Discouragement is a mechanism to prevent our limited connection slots being used up by incompatible or broken peers.

So I think we should just set `fDisconnect` for any peer that sends us a `FILTER*` message when we don't have `NODE_BLOOM` enabled, regardless of their version number.

MarcoFalke replied:
In theory it is possible to set `-banscore=101` to not disconnect a 100-misbehaving peer. Though, that would only kick down the can by one more message. I can't see a use case for that and encouraging users to change banscore to accomodate misbehaving light clients seems ill-advised. So I agree with the suggestion made by @jnewbery .

jnewbery replied:
We should remove `-banscore` entirely. It's not useful user configuration.

gzhao408 replied:
Thanks for the suggestion! I've updated it but I think Travis timed out on one of the builds.

jnewbery commented:
Code review ACK 18a2e0a3f7071b8df6b20ecc1b34b3f2abb13cfc

Great tidy-up. Thanks Gloria!

naumenkogs commented:
utACK 18a2e0a

I was somewhat confused by the last commit, because the description of the PR is about `filterclear`, but the last commit has absolutely nothing to do with filterclear.
Not sure if we want to do anything about it :)

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3490,13 +3490,15 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
     if (msg_type == NetMsgType::FILTERCLEAR) {
+        if (!(pfrom.GetLocalServices() & NODE_BLOOM)) {
+            pfrom.fDisconnect = true;
+            return false;
```
in commit 294b96772b7d184a0cd6f56bd76ef118f4c78cc6:

Any reason to return false here? The only effect is that it will log a line with `FAILED` with the NET debug category. However, it should be pretty clear from the log already that receiving a filterclear message and then disconnecting a peer means the message was ignored.

naumenkogs replied:
I was confused here too, but it seems like this is a common thing to do. Many occurences of 
``
             pfrom.fDisconnect = true;
            return false; 
``

MarcoFalke replied:
The majority of disconnects is followed by a `return true;`. The ones that don't should be changed too. And longer term the return value should be removed completely.

gzhao408 replied:
Thanks! Yup, my reasoning was that's what it looked like in other places 😅

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3466,6 +3456,10 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
     if (msg_type == NetMsgType::FILTERADD) {
+        if (!(pfrom.GetLocalServices() & NODE_BLOOM)) {
+            pfrom.fDisconnect = true;
+            return false;
```
in commit 18a2e0a3f7071b8df6b20ecc1b34b3f2abb13cfc:

Same

maflcko reviewed (approved):
ACK 18a2e0a3f7071b8df6b20ecc1b34b3f2abb13cfc nice cleanup 🌫

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 18a2e0a3f7071b8df6b20ecc1b34b3f2abb13cfc nice cleanup 🌫
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhzKAv/ejpwEnqSww3TRDRR7fvVaVti9v81jtyV0QVIwfjHvYw3MD3BSifUI8KF
KCWYTX/MVDobco2GQR3Iw1Xnc1oGgH8416c5ehDMTiO9T7+DftefLR1hHb8nidXs
bPrsTfbwvh7TkPsEwVWjSu+lVL1CtYDHDcdyP/t2u8p/r/RpCq1WxEwmSQgqybTK
rCIdWUNQ1yrNd5/gkpydpzIZPLpui19aA+lxwndwuWiME0hTc2li/qE9acsFSLSI
PXM/gClEJ6aKDyVX2AOGfpQv1KEY5OReDsUBi2jdXRXI7n3/1D6kz9YGar4DIhVy
nuVcQFEEMGvGEclQVeIp5ik5JHPta0bFm/i6rSdHSk1Y+b5zqgYmSeNqnUwcUxcG
5jyVXszUw0p61yVgy/7CFaFvIr8oa2xjAt8vvLogdePIjH320j2TA3UBhOWLBkfU
bcWafRJwm13cB3VtSB00HNZ7LBb5mS8AmsV0sY4U4a9//RbgBv/HHgrTkjbFmiCQ
gqH6/zVZ
=Wizi
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `76972080cdb7d6db860684d9cc7ddc281674c2c67368d5f83cad36c0fa0849c0  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010876972080cdb7d6db860684d9cc7ddc281674c2c67368d5f83cad36c0fa0849c0f0109dbec04e71727af86b26a69e0456ceed08fff0103e0ccf025069a00a115eb36f0239bf9908f1045ee6212ef008406edf3aafd50e7b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105d13e0631ca2fc0dffb94eb69c9e257e08f1045ee6212ff008ec3497fbdc12247d0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f0108ee3c4887c4a6fb0178cf83c0894e49f08f1045ee62130f0082f9de07408937afb0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

MarcoFalke commented:
re-ACK 3a10d935ac only change is replacing false with true 🚝

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 3a10d935ac only change is replacing false with true 🚝
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiicAwAuImcvIy26lnaw4VoTkzdS0iEgf2zA6xyuFKb3zUCSKpX80mPbnrdDxue
32BIzcZfgyy5FvhWuJddItl3wEdmnyJK/1vZqRogs/Q4oljfVaDxzdIPKc3lhoG/
GNOut9oyn0WAhL26QQDdqWwCq94GFJ7DSClIfsozyXQcs61ylKQULEBpP039uR1R
/PQEKZoS8t2ThDQXHgcLmC04QYARp8PX5RvSBbxFQz75Kl9vUD9Qfs3/KlCHBThU
dKwWnLt/yQp3w0Wz4XVgjborfhgeLVVzmvkBRZboFEMXJPaC1OkinxS3nydh6MYz
7fMTVM61b95qy21pBSr0iT8T/gslbr7sj2H9OEkfU7t8s096qnPfVgd/qTtMmneR
v5NN9WVf9SXq8nmhGZQPLL/gehW1RGC99vSL3MwrQQwspz13lU+3wAXEGV16k7Py
uFKUrHkUlNitTxn9vFWoeTLIlWFtJ6Qc2B6dc3wMJ7k1KHfv2yI1CA1OcdO4osNv
pRAS6UPd
=ZuJt
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `d49fb833782cb07ba218eb5ea866b2c2f12e58a1ac6670b3a225942fdd3a8e52  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108d49fb833782cb07ba218eb5ea866b2c2f12e58a1ac6670b3a225942fdd3a8e52f0103c04b8c6d8a1b725bb8031c53f57b9d108fff0107c17668fb0d82be83eef6731e41506b808f1045ee7547cf008a80c1881e09b04be0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010ace2c1ce54b81da5a48e5bebd5a4385108f0208ea4598e8802f435f56f7d8c101decd080a2814c8962b6b28136630218f9926908f1045ee7547ef008cecf3530c651d82e0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010bba6ccc6d73c82c96b9749ba67d7925a08f1045ee7547cf0081c33a0e9f53e4e320083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jnewbery commented:
Code review ACK 3a10d935ac8ebabdfd336569d943f042ff84b13e

gillichu commented:
Concept ACK, I just learned a lot about how bloom filters are used in Bitcoin.

gillichu commented:
tested ACK: just did a quick run of test_runner with the macOS

gillichu commented:
tested ACK: quick test_runner on macOS [3a10d93](https://github.com/bitcoin/bitcoin/commit/3a10d935ac8ebabdfd336569d943f042ff84b13e)

gzhao408 commented:
😂  Thanks @gillichu, I really appreciate the effort 🙏  

It's \[\`commit\`\]\(link\)

jnewbery commented:
> It's \[\`commit\`\]\(link\)

No need to use markdown for links to the commit hash. Just paste the full hash and github will automatically convert it to a link to commit.

gzhao408 commented:
> Just paste the full hash and github will automatically convert it to a link to commit.

WHAT I've been using markdown this whole time 😭

naumenkogs commented:
utACK 3a10d93

rebroad commented:
> Concept ACK. There shouldn't be a reason for a node to send filterclear when we haven't offered them bloom services to them.
> 
> See also BIP 111:
> 
> > If a node does not support bloom filters but receives a "filterload", "filteradd", or "filterclear" message from a peer the node should disconnect that peer immediately
> 
> https://github.com/bitcoin/bips/blob/master/bip-0111.mediawiki
> 
> nit: I think this is a good opportunity to move the disconnect handling of `NetMsgType::FILTERLOAD` and ` NetMsgType::FILTERADD` into the respective scope of the switch statement instead of having it at the top of the function, disconnected from the actual filter message handling.

filterclear was also available to request a node sends TXs after IBD had completed - this worked better than using feefilter as it was part of the connection handshake, unlike feefilter.

MarcoFalke commented:
@rebroad Please, there is no reason to leave comments on three pull requests to ask questions. The thread you created in #21839 is sufficient.
